core/src/main/scala/com/microsoft/azure/synapse/ml/causal/LinearDMLEstimator.scala

```diff
@@ -0,0 +1,335 @@
+    val train = splits(0).cache()
+    val test = splits(1).cache()
+
+    // Step 2 - use first sample data to fit treatment model and outcome model
+    val (treatmentEstimator, treatmentResidualPredictionColName, treatmentPredictionColsToDrop) = getTreatmentModel match {
+      case classifier: ProbabilisticClassifier[_, _, _] => (
+        new TrainClassifier()
+          .setFeaturesCol("treatment_features")
+          .setModel(getTreatmentModel)
+          .setLabelCol(getTreatmentCol)
+          .setExcludedFeatures(Array(getOutcomeCol)),
+        classifier.getProbabilityCol,
+        Seq(classifier.getPredictionCol, classifier.getProbabilityCol, classifier.getRawPredictionCol)
+      )
+      case regressor: Regressor[_, _, _] => (
+        new TrainRegressor()
+          .setFeaturesCol("treatment_features")
+          .setModel(getTreatmentModel)
+          .setLabelCol(getTreatmentCol)
+          .setExcludedFeatures(Array(getOutcomeCol)),
+        regressor.getPredictionCol,
+        Seq(regressor.getPredictionCol)
+      )
+    }
+
+    val (outcomeEstimator, outcomeResidualPredictionColName, outcomePredictionColsToDrop) = getOutcomeModel match {
+      case classifier: ProbabilisticClassifier[_, _, _] => (
+        new TrainClassifier()
+          .setFeaturesCol("outcome_features")
+          .setModel(getOutcomeModel)
+          .setLabelCol(getOutcomeCol)
+          .setExcludedFeatures(Array(getTreatmentCol)),
+        classifier.getProbabilityCol,
+        Seq(classifier.getPredictionCol, classifier.getProbabilityCol, classifier.getRawPredictionCol)
+      )
+      case regressor: Regressor[_, _, _] => (
+        new TrainRegressor()
+          .setFeaturesCol("outcome_features")
+          .setModel(getOutcomeModel)
+          .setLabelCol(getOutcomeCol)
+          .setExcludedFeatures(Array(getTreatmentCol)),
+        regressor.getPredictionCol,
+        Seq(regressor.getPredictionCol)
+      )
+    }
+
+    val treatmentModelV1 = treatmentEstimator.fit(train)
+    val outcomeModelV1 = outcomeEstimator.fit(train)
+
+    // Step 3 - use second sample data to get predictions and compute residuals
+    val treatmentPredictedDFV1 = treatmentModelV1.transform(test)
+    val treatmentResidualDFV1 =
+      new ComputeResidualTransformer()
+        .setObservedCol(getTreatmentCol)
+        .setPredictedCol(treatmentResidualPredictionColName)
+        .setOutcomeCol(SchemaConstants.TreatmentResidualColumn)
+        .transform(treatmentPredictedDFV1)
+        .drop(treatmentPredictionColsToDrop: _*)
+
+    val outcomePredictedDFV1 = outcomeModelV1.transform(treatmentResidualDFV1)
+
+    val residualsDFV1 =
+      new ComputeResidualTransformer()
+        .setObservedCol(getOutcomeCol)
+        .setPredictedCol(outcomeResidualPredictionColName)
+        .setOutcomeCol(SchemaConstants.OutcomeResidualColumn)
+        .transform(outcomePredictedDFV1)
+        .drop(outcomePredictionColsToDrop: _*)
```

Comment:
Consider combining the two residual transformers into a pipeline and do it together.

Reply:
Also this part of the code is duplicated below. Consider extracting it to a function and reuse during cross fitting.

Reply:
I didn't put them into a pipeline due to a bug https://github.com/microsoft/SynapseML/issues/1692
and it was fixed recently, let me try move them to one pipeline.

Reply:
updated code to use pipeline for these steps